sos/collector/clusters/pacemaker.py

```diff
@@ -18,42 +22,80 @@ class pacemaker(Cluster):
-                offline.extend([n for n in nodes.split(' ') if n])
-        return offline
+        self.nodes = []
+        # try crm_mon first
+        try:
+            if not self.get_option('only-corosync'):
+                try:
+                    self.get_nodes_from_crm()
+                except Exception as err:
+                    self.log_warn("Could not parse crm_mon output: %s" % err)
+            if not self.nodes:
+                # fallback to corosync.conf, in case the node we're inspecting
+                # is offline from the cluster
+                self.get_nodes_from_corosync()
+        except Exception as err:
+            self.log_error("Could not determine nodes from cluster: %s" % err)
+
+        _shorts = [n for n in self.nodes if '.' not in n]
+        if _shorts:
+            self.log_warn(
+                "WARNING: Node addresses '%s' may not resolve locally if you "
+                "are not running on a node in the cluster. Try using option "
+                "'-c pacemaker.only-corosync' if these connections fail."
+                % ','.join(_shorts)
+            )
+        return self.nodes
+
+    def get_nodes_from_crm(self):
+        """
+        Try to parse crm_mon output for node list and status.
+        """
+        xmlopt = '--output-as=xml'
+        # older pacemaker had a different option for xml output
+        _ver = self.exec_primary_cmd('crm_mon --version')
+        if _ver['status'] == 0:
+            cver = _ver['output'].split()[1].split('-')[0]
+            if not version.parse(cver) > version.parse('2.0.3'):
+                xmlopt = '--as-xml'
+        else:
+            return
+        _out = self.exec_primary_cmd(
+            "crm_mon --one-shot --inactive %s" % xmlopt,
+            need_root=True
+        )
+        if _out['status'] == 0:
+            self.parse_crm_xml(_out['output'])
+
+    def parse_crm_xml(self, xmlstring):
+        """
+        Parse the xml output string provided by crm_mon
+        """
+        _xml = ElementTree.fromstring(xmlstring)
+        nodes = _xml.find('nodes')
+        for node in nodes:
+            _node = node.attrib
+            if self.get_option('online') and _node['online'] == 'true':
+                self.nodes.append(_node['name'])
+            elif self.get_option('offline') and _node['online'] == 'false':
+                self.nodes.append(_node['name'])
+
+    def get_nodes_from_corosync(self):
+        """
+        As a fallback measure, read corosync.conf to get the node list. Note
+        that this prevents us from separating online nodes from offline nodes.
+        """
+        self.log_warn("WARNING: using corosync.conf as a fallback measure to "
+                      "enumerate nodes. This method is unable to separate "
+                      "online nodes from offline nodes.")
+        cc = self.primary.read_file('/etc/corosync/corosync.conf')
+        nodes = re.findall(r'((\sring0_addr:)(.*))', cc)
```

Comment:
These should be resolvable (from **within** the cluster), but may not be the node names from pacemaker's perspective. They may not be names at all -- they could be IP addresses. If that doesn't create any issues for the collection or resulting report bundle (i.e., if we're using these addrs only to connect to each target node), then that's fine. I haven't had a chance to test.

Reply:
IP addresses are fine. We use these to make the connections to the nodes and in our initial logging for those connections only.